src/udev/udevadm-settle.c

```diff
@@ -87,54 +89,69 @@ static int parse_argv(int argc, char *argv[]) {
-        if (r < 0 || !streq(unit, "systemd-udev-settle.service"))
+        if (r < 0) {
+                log_debug_errno(r, "Failed to determine unit we run in, ignoring: %m");
                 return 0;
+        }
 
-        log_notice("systemd-udev-settle.service is deprecated.");
+        if (!streq(unit, "systemd-udev-settle.service"))
+                return 0;
 
-        r = sd_bus_open_system(&bus);
-        if (r < 0)
-                return log_debug_errno(r, "Failed to open system bus, skipping dependency queries: %m");
-
-        unit_path = unit_dbus_path_from_name("systemd-udev-settle.service");
-        if (!unit_path)
-                return -ENOMEM;
-
-        (void) sd_bus_get_property_strv(
-                        bus,
-                        "org.freedesktop.systemd1",
-                        unit_path,
-                        "org.freedesktop.systemd1.Unit",
-                        "WantedBy",
-                        NULL,
-                        &a);
-
-        (void) sd_bus_get_property_strv(
-                        bus,
-                        "org.freedesktop.systemd1",
-                        unit_path,
-                        "org.freedesktop.systemd1.Unit",
-                        "RequiredBy",
-                        NULL,
-                        &b);
-
-        r = strv_extend_strv(&a, b, true);
+        r = bus_connect_system_systemd(&bus);
         if (r < 0)
-                return r;
+                log_debug_errno(r, "Failed to open connection to systemd, skipping dependency queries: %m");
+        else {
+                _cleanup_strv_free_ char **b = NULL;
+                _cleanup_free_ char *unit_path = NULL;
+
+                unit_path = unit_dbus_path_from_name("systemd-udev-settle.service");
+                if (!unit_path)
+                        return -ENOMEM;
+
+                (void) sd_bus_get_property_strv(
+                                bus,
+                                "org.freedesktop.systemd1",
+                                unit_path,
+                                "org.freedesktop.systemd1.Unit",
+                                "WantedBy",
+                                NULL,
+                                &a);
+
+                (void) sd_bus_get_property_strv(
+                                bus,
+                                "org.freedesktop.systemd1",
+                                unit_path,
+                                "org.freedesktop.systemd1.Unit",
+                                "RequiredBy",
+                                NULL,
+                                &b);
+
+                r = strv_extend_strv(&a, b, true);
+                if (r < 0)
+                        return r;
+        }
 
-        if (!strv_isempty(a)) {
+        if (strv_isempty(a))
+                log_struct(LOG_NOTICE,
+                           "MESSAGE=systemd-udev-settle.service is deprecated.",
+                           "OFFENDING_UNITS=none",
```

Comment:
I don't think it's useful to log the message with OFFENDING_UNITS=none. I don't see a way to make the message in the catalog not confusing in this case, since it talks about reporting bugs to those other units. People call `udevadm settle` from scripts and there it reasonable to use it. So I think if branch of the conditional should be removed again.

Reply:
The only reason I added "none" here is that we have no good story on what to do with catalog entries that include `@@` references to variables that do not exist in the log entry...

it's less confusing to show "none" then an empty string here...

Maybe we should slightly extend the catalog template language, and maybe support some really minimal if/else logic that can check for variables being defined. i.e.

```
#ifdef OFFENDING_UNITS
Version 1 of the paragraph
#else
Version 2 of the paragraph
#endif
```

Taking inspiration from C syntax...

with that in place we can handle that much nicer. But until then I think we should keep the "none" thing, wouldn't you agree?

Reply:
Posted #16449 requesting this

Reply:
>  People call udevadm settle from scripts and there it reasonable to use it. So I think if branch of the conditional should be removed again.

The message is not generated outside of systemd-udev-settle.service actually (i.e. the code checks in which unit it runs — you wrote that code btw ;-)). This this should never trigger in scripts hence, as they run in other units...

Reply:
I solved this now by generating a simple log_notice() message if we can#t disocver the deps. Thus you'll get the catalog stuff only if we know the deps.